File: src/circuit-breaker.ts
/**
 * Circuit Breaker Pattern Implementation
 *
 * States:
 * - CLOSED: Normal operation, requests go to primary service (Anthropic)
 * - OPEN: Service detected as down, requests go to fallback (LMStudio)
 * - HALF_OPEN: Testing if primary service has recovered
 *
 * Based on resilience best practices:
 * - Fast failure detection with timeouts
 * - Automatic recovery testing
 * - Configurable thresholds
 */

export enum CircuitState {
  CLOSED = "CLOSED", // Normal - using Anthropic
  OPEN = "OPEN", // Failover - using LMStudio
  HALF_OPEN = "HALF_OPEN", // Testing - trying Anthropic again
}

export interface CircuitBreakerConfig {
  failureThreshold: number; // Failures before opening circuit
  successThreshold: number; // Successes in HALF_OPEN before closing
  retryTimeout: number; // Time before trying HALF_OPEN (ms)
  requestTimeout: number; // Max request time before considering failed (ms)
  latencyThresholdMs?: number; // Latency threshold for circuit breaking (0 = disabled)
  latencyConsecutiveChecks?: number; // Number of consecutive high latency checks before opening
  latencyWindowMs?: number; // Rolling window for latency samples (ms)
  autoCheckLatency?: boolean; // Automatically check latency threshold on recordLatency
  maxLatencySamples?: number; // Max samples to keep (prevents unbounded growth - DoS protection)
}

// Security: Maximum latency samples to prevent DoS via memory exhaustion
const DEFAULT_MAX_LATENCY_SAMPLES = 10000;

export interface CircuitBreakerMetrics {
  state: CircuitState;
  failureCount: number;
  successCount: number;
  avgLatencyMs: number;
  latencySamples: number;
  minLatencyMs: number;
  maxLatencyMs: number;
  p50LatencyMs: number;
  p95LatencyMs: number;
  p99LatencyMs: number;
  consecutiveHighLatency: number;
  nextAttempt: string | null;
  timestamp: string;
}

interface LatencySample {
  timestamp: number;
  latencyMs: number;
}

export class CircuitBreaker {
  private state: CircuitState = CircuitState.CLOSED;
  private failureCount: number = 0; // For state machine logic
  private successCount: number = 0; // For state machine logic
  private totalFailureCount: number = 0; // For metrics
  private totalSuccessCount: number = 0; // For metrics
  private nextAttempt: number = Date.now();
  private config: CircuitBreakerConfig;
  private onStateChange?: (state: CircuitState, reason?: string) => void;
  private latencySamples: LatencySample[] = [];
  private consecutiveHighLatency: number = 0;

  constructor(config: Partial<CircuitBreakerConfig> = {}) {
    this.config = {
      failureThreshold: config.failureThreshold ?? 3,
      successThreshold: config.successThreshold ?? 2,
      retryTimeout: config.retryTimeout ?? 30000, // 30 seconds
      requestTimeout: config.requestTimeout ?? 5000, // 5 seconds
      latencyThresholdMs: config.latencyThresholdMs ?? 0, // Disabled by default
      latencyConsecutiveChecks: config.latencyConsecutiveChecks ?? 3,
      latencyWindowMs: config.latencyWindowMs ?? 1000, // 1 second window
      autoCheckLatency: config.autoCheckLatency ?? false,
      maxLatencySamples:
        config.maxLatencySamples ?? DEFAULT_MAX_LATENCY_SAMPLES,
    };
  }

  /**
   * Register callback for state changes (for logging/monitoring)
   */
  public onStateChangeListener(
    callback: (state: CircuitState, reason?: string) => void
  ): void {
    this.onStateChange = callback;
  }

  /**
   * Check if request should be allowed to primary service
   */
  public shouldAllowRequest(): boolean {
    if (this.state === CircuitState.CLOSED) {
      return true;
    }

    if (this.state === CircuitState.OPEN) {
      // Check if it's time to try HALF_OPEN
      if (Date.now() >= this.nextAttempt) {
        this.setState(CircuitState.HALF_OPEN);
        return true;
      }
      return false; // Use fallback
    }

    // HALF_OPEN state - allow request to test recovery
    return true;
  }

  /**
   * Record successful request
   */
  public recordSuccess(): void {
    this.failureCount = 0;
    this.totalSuccessCount++; // Track for metrics

    if (this.state === CircuitState.HALF_OPEN) {
      this.successCount++;
      if (this.successCount >= this.config.successThreshold) {
        this.setState(CircuitState.CLOSED);
        this.successCount = 0;
      }
    }
  }

  /**
   * Record failed request
   */
  public recordFailure(): void {
    this.failureCount++;
    this.totalFailureCount++; // Track for metrics
    this.successCount = 0;

    if (this.state === CircuitState.HALF_OPEN) {
      // Failed during recovery test - go back to OPEN
      this.setState(CircuitState.OPEN);
      this.scheduleRetry();
    } else if (this.failureCount >= this.config.failureThreshold) {
      // Too many failures - open circuit
      this.setState(CircuitState.OPEN);
      this.scheduleRetry();
    }
  }

  /**
   * Manually trip the circuit (for testing)
   */
  public trip(): void {
    this.setState(CircuitState.OPEN);
    this.scheduleRetry();
  }

  /**
   * Manually reset the circuit (for testing)
   */
  public reset(): void {
    this.failureCount = 0;
    this.successCount = 0;
    this.totalFailureCount = 0;
    this.totalSuccessCount = 0;
    this.latencySamples = [];
    this.consecutiveHighLatency = 0;
    this.setState(CircuitState.CLOSED);
  }

  /**
   * Record latency measurement
   * Security: Validates input to prevent NaN/Infinity poisoning and enforces sample limits
   */
  public recordLatency(latencyMs: number): void {
    // Security: Complete input validation (CWE-20)
    if (latencyMs <= 0) {
      throw new Error("Latency must be positive");
    }
    if (!Number.isFinite(latencyMs)) {
      throw new Error("Latency must be a finite number");
    }
    if (latencyMs > 86400000) {
      // 24 hours max - sanity check
      throw new Error("Latency exceeds maximum allowed value (24 hours)");
    }

    // Add sample with timestamp
    this.latencySamples.push({
      timestamp: Date.now(),
      latencyMs,
    });

    // Security: Enforce max samples limit to prevent DoS via memory exhaustion (CWE-400)
    if (this.latencySamples.length > this.config.maxLatencySamples!) {
      // Remove oldest samples to stay within limit
      const excess =
        this.latencySamples.length - this.config.maxLatencySamples!;
      this.latencySamples.splice(0, excess);
    }

    // Auto-check latency threshold if enabled
    if (this.config.autoCheckLatency) {
      this.checkLatencyThreshold();
    }
  }

  /**
   * Check if average latency exceeds threshold
   */
  public checkLatencyThreshold(): void {
    // Skip if threshold disabled
    if (
      !this.config.latencyThresholdMs ||
      this.config.latencyThresholdMs === 0
    ) {
      return;
    }

    const validSamples = this.getValidLatencySamples();

    // Need samples to check
    if (validSamples.length === 0) {
      this.consecutiveHighLatency = 0;
      return;
    }

    // Get the most recent sample
    const lastSample = validSamples[validSamples.length - 1];

    // Check if last sample is above threshold (for consecutive tracking)
    if (lastSample.latencyMs >= this.config.latencyThresholdMs!) {
      this.consecutiveHighLatency++;

      // Open circuit if threshold reached
      if (
        this.consecutiveHighLatency >= this.config.latencyConsecutiveChecks!
      ) {
        this.setState(CircuitState.OPEN, "latency threshold exceeded");
        this.scheduleRetry();
      }
    } else {
      // Reset counter on normal latency
      this.consecutiveHighLatency = 0;
    }
  }

  /**
   * Get complete metrics
   */
  public getMetrics(): CircuitBreakerMetrics {
    const validSamples = this.getValidLatencySamples();
    const latencies = validSamples.map((s) => s.latencyMs);

    // Calculate basic stats - round to 2 decimal places
    const rawAvg =
      latencies.length > 0
        ? latencies.reduce((sum, lat) => sum + lat, 0) / latencies.length
        : 0;
    const avgLatencyMs = Math.round(rawAvg * 100) / 100;
    const minLatencyMs = latencies.length > 0 ? Math.min(...latencies) : 0;
    const maxLatencyMs = latencies.length > 0 ? Math.max(...latencies) : 0;

    // Calculate percentiles
    const p50LatencyMs =
      Math.round(this.calculatePercentile(latencies, 50) * 100) / 100;
    const p95LatencyMs =
      Math.round(this.calculatePercentile(latencies, 95) * 100) / 100;
    const p99LatencyMs =
      Math.round(this.calculatePercentile(latencies, 99) * 100) / 100;

    return {
      state: this.state,
      failureCount: this.totalFailureCount,
      successCount: this.totalSuccessCount,
      avgLatencyMs,
      latencySamples: validSamples.length,
      minLatencyMs,
      maxLatencyMs,
      p50LatencyMs,
      p95LatencyMs,
      p99LatencyMs,
      consecutiveHighLatency: this.consecutiveHighLatency,
      nextAttempt:
        this.state === CircuitState.OPEN
          ? new Date(this.nextAttempt).toISOString()
          : null,
      timestamp: new Date().toISOString(),
    };
  }

  /**
   * Static HTTP endpoint handler for metrics
   * Security: No CORS headers - this is an internal monitoring endpoint
   * Authentication should be handled at the proxy/gateway level
   */
  public static async handleMetricsRequest(
    breaker: CircuitBreaker,
    req: { method: string; url: string },
    res: {
      writeHead: (status: number, headers: Record<string, string>) => void;
      end: (body?: string) => void;
    }
  ): Promise<void> {
    // Check method and URL
    if (req.method !== "GET" || req.url !== "/v1/circuit-breaker/metrics") {
      res.writeHead(404, { "Content-Type": "text/plain" });
      res.end("Not Found");
      return;
    }

    // Get metrics
    const metrics = breaker.getMetrics();

    // Return JSON response (no CORS - internal endpoint only)
    res.writeHead(200, {
      "Content-Type": "application/json",
    });
    res.end(JSON.stringify(metrics));
  }

  /**
   * Get valid latency samples within the rolling window
   */
  private getValidLatencySamples(): LatencySample[] {
    const now = Date.now();
    const cutoff = now - this.config.latencyWindowMs!;

    // Filter samples within window (strictly greater than cutoff to exclude boundary)
    return this.latencySamples.filter((sample) => sample.timestamp > cutoff);
  }

  /**
   * Calculate percentile from latency array using linear interpolation
   */
  private calculatePercentile(latencies: number[], percentile: number): number {
    if (latencies.length === 0) {
      return 0;
    }

    // Sort latencies
    const sorted = [...latencies].sort((a, b) => a - b);

    // For single element
    if (sorted.length === 1) {
      return sorted[0];
    }

    // Calculate index using linear interpolation between ranks
    // This gives more accurate percentiles for small datasets
    const index = (percentile / 100) * (sorted.length - 1);
    const lower = Math.floor(index);
    const upper = Math.ceil(index);
    const weight = index - lower;

    // Interpolate between lower and upper values
    if (lower === upper) {
      return sorted[lower];
    }

    return sorted[lower] * (1 - weight) + sorted[upper] * weight;
  }

  /**
   * Get current state
   */
  public getState(): CircuitState {
    return this.state;
  }

  /**
   * Get request timeout from config
   */
  public getRequestTimeout(): number {
    return this.config.requestTimeout;
  }

  /**
   * Get statistics for monitoring
   */
  public getStats() {
    return {
      state: this.state,
      failureCount: this.failureCount,
      successCount: this.successCount,
      nextAttempt:
        this.state === CircuitState.OPEN
          ? new Date(this.nextAttempt).toISOString()
          : null,
    };
  }

  private setState(newState: CircuitState, reason?: string): void {
    if (this.state !== newState) {
      this.state = newState;

      // Reset metrics when circuit closes
      if (newState === CircuitState.CLOSED) {
        this.consecutiveHighLatency = 0;
        this.totalFailureCount = 0; // Reset failure count on recovery
      }

      this.onStateChange?.(newState, reason);
    }
  }

  private scheduleRetry(): void {
    this.nextAttempt = Date.now() + this.config.retryTimeout;
  }
}

File: scripts/lib/circuit_breaker.py
#!/usr/bin/env python3
"""
Circuit Breaker Pattern

Implements circuit breaker pattern to protect against cascading failures.
Provides three states: CLOSED (normal), OPEN (failing), HALF_OPEN (testing recovery).

Classes:
    - CircuitBreakerState: Enum for circuit states
    - CircuitBreakerError: Exception when circuit is open
    - CircuitBreakerMetrics: Metrics data class
    - CircuitBreaker: Main circuit breaker implementation
"""

import time
import threading
from enum import Enum
from typing import Callable, Any, Optional, Dict, List
from dataclasses import dataclass, field


class CircuitBreakerState(Enum):
    """Circuit breaker states"""
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


class CircuitBreakerError(Exception):
    """Raised when circuit breaker rejects a request"""
    pass


@dataclass
class CircuitBreakerMetrics:
    """Circuit breaker metrics"""
    total_calls: int = 0
    successes: int = 0
    failures: int = 0
    rejections: int = 0
    state_changes: List[Dict[str, Any]] = field(default_factory=list)

    @property
    def failure_rate(self) -> float:
        """Calculate failure rate"""
        if self.total_calls == 0:
            return 0.0
        return self.failures / self.total_calls

    @property
    def rejection_rate(self) -> float:
        """Calculate rejection rate"""
        if self.total_calls == 0:
            return 0.0
        return self.rejections / self.total_calls


class CircuitBreaker:
    """
    Circuit breaker for protecting against cascading failures

    State Machine:
    - CLOSED: Normal operation, track failures
    - OPEN: Reject all requests, check recovery timeout
    - HALF_OPEN: Allow test requests, close on success or reopen on failure

    Thread-safe for concurrent operations.
    """

    def __init__(
        self,
        failure_threshold: int = 5,
        recovery_timeout: float = 60.0,
        success_threshold: int = 2,
        name: str = "circuit_breaker"
    ):
        """
        Initialize circuit breaker

        Args:
            failure_threshold: Number of failures before opening circuit
            recovery_timeout: Seconds to wait before attempting recovery
            success_threshold: Successes needed in HALF_OPEN to close circuit
            name: Circuit breaker identifier
        """
        self.failure_threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.success_threshold = success_threshold
        self.name = name

        # State tracking
        self._state = CircuitBreakerState.CLOSED
        self._failure_count = 0
        self._success_count = 0
        self._last_failure_time = None
        self._lock = threading.Lock()

        # Metrics
        self._metrics = CircuitBreakerMetrics()

    @property
    def state(self) -> CircuitBreakerState:
        """Get current state (thread-safe)"""
        with self._lock:
            return self._state

    def call(self, func: Callable, *args, **kwargs) -> Any:
        """
        Execute function with circuit breaker protection

        Args:
            func: Function to execute
            *args: Function arguments
            **kwargs: Function keyword arguments

        Returns:
            Function result

        Raises:
            CircuitBreakerError: If circuit is open
            Exception: If function raises exception
        """
        with self._lock:
            # Check if we should attempt recovery
            if self._state == CircuitBreakerState.OPEN:
                if self._should_attempt_reset():
                    self._transition_to(CircuitBreakerState.HALF_OPEN)
                else:
                    # Still in open state, reject request
                    self._metrics.total_calls += 1
                    self._metrics.rejections += 1
                    raise CircuitBreakerError(
                        f"Circuit breaker '{self.name}' is OPEN"
                    )

            # Record call attempt
            self._metrics.total_calls += 1

        # Execute function (outside lock to prevent blocking)
        try:
            result = func(*args, **kwargs)
            self._on_success()
            return result
        except Exception as e:
            self._on_failure()
            raise

    def get_state(self) -> CircuitBreakerState:
        """Get current state"""
        return self.state

    def get_metrics(self) -> CircuitBreakerMetrics:
        """Get circuit breaker metrics"""
        with self._lock:
            # Return copy to prevent external modification
            return CircuitBreakerMetrics(
                total_calls=self._metrics.total_calls,
                successes=self._metrics.successes,
                failures=self._metrics.failures,
                rejections=self._metrics.rejections,
                state_changes=self._metrics.state_changes.copy()
            )

    def reset(self) -> None:
        """Manually reset circuit breaker to CLOSED state"""
        with self._lock:
            self._transition_to(CircuitBreakerState.CLOSED)
            self._failure_count = 0
            self._success_count = 0
            self._last_failure_time = None

    def _on_success(self) -> None:
        """Handle successful function call"""
        with self._lock:
            self._metrics.successes += 1

            if self._state == CircuitBreakerState.HALF_OPEN:
                # Increment success count in HALF_OPEN
                self._success_count += 1

                # Check if we've reached success threshold
                if self._success_count >= self.success_threshold:
                    self._transition_to(CircuitBreakerState.CLOSED)
                    self._failure_count = 0
                    self._success_count = 0

            elif self._state == CircuitBreakerState.CLOSED:
                # Reset failure count on success in CLOSED state
                self._failure_count = 0

    def _on_failure(self) -> None:
        """Handle failed function call"""
        with self._lock:
            self._metrics.failures += 1
            self._failure_count += 1
            self._last_failure_time = time.time()

            if self._state == CircuitBreakerState.CLOSED:
                # Check if we've reached failure threshold
                if self._failure_count >= self.failure_threshold:
                    self._transition_to(CircuitBreakerState.OPEN)

            elif self._state == CircuitBreakerState.HALF_OPEN:
                # Failure in HALF_OPEN reopens circuit
                self._transition_to(CircuitBreakerState.OPEN)
                self._success_count = 0

    def _should_attempt_reset(self) -> bool:
        """Check if recovery timeout has elapsed"""
        if self._last_failure_time is None:
            return False

        elapsed = time.time() - self._last_failure_time
        return elapsed >= self.recovery_timeout

    def _transition_to(self, new_state: CircuitBreakerState) -> None:
        """
        Transition to new state and record change

        Must be called within lock.
        """
        old_state = self._state
        self._state = new_state

        # Record state change (bounded to prevent memory leak)
        MAX_STATE_CHANGES = 1000
        self._metrics.state_changes.append({
            'from_state': old_state.value,
            'to_state': new_state.value,
            'timestamp': time.time()
        })

        # Keep only most recent changes
        if len(self._metrics.state_changes) > MAX_STATE_CHANGES:
            self._metrics.state_changes = self._metrics.state_changes[-MAX_STATE_CHANGES:]
